zookeeper-api/src/test/java/org/apache/helix/zookeeper/util/TestHttpRoutingDataReader.java

```diff
@@ -0,0 +1,124 @@
+    _testRawRoutingData
+        .put("zk-1", ImmutableSet.of("/sharding-key-3", "/sharding-key-4", "/sharding-key-5"));
+    _testRawRoutingData
+        .put("zk-2", ImmutableSet.of("/sharding-key-6", "/sharding-key-7", "/sharding-key-8"));
+
+    // Start MockMSDS
+    _msdsServer =
+        new MockMetadataStoreDirectoryServer(_host, _port, _namespace, _testRawRoutingData);
+    _msdsServer.startServer();
+
+    // Register the endpoint as a System property
+    String msdsEndpoint = "http://" + _host + ":" + _port + "/admin/v2/namespaces/" + _namespace;
+    System.setProperty(MetadataStoreRoutingConstants.MSDS_SERVER_ENDPOINT_KEY, msdsEndpoint);
+  }
+
+  @AfterClass
+  public void afterClass() {
+    _msdsServer.stopServer();
+  }
+
+  @Test
+  public void testGetRawRoutingData() throws IOException {
+    Map<String, List<String>> rawRoutingData = HttpRoutingDataReader.getRawRoutingData();
+    _testRawRoutingData
+        .forEach((realm, keys) -> Assert.assertEquals(rawRoutingData.get(realm), keys));
+  }
+
+  @Test(dependsOnMethods = "testGetRawRoutingData")
+  public void testGetMetadataStoreRoutingData() throws IOException, InvalidRoutingDataException {
+    MetadataStoreRoutingData data = HttpRoutingDataReader.getMetadataStoreRoutingData();
+    Map<String, String> allMappings = data.getAllMappingUnderPath("/");
+    Map<String, Set<String>> groupedMappings = allMappings.entrySet().stream().collect(Collectors
+        .groupingBy(Map.Entry::getValue,
+            Collectors.mapping(Map.Entry::getKey, Collectors.toSet())));
+    _testRawRoutingData.forEach((realm, keys) -> {
+      // Two way containsAll because AssertEquals on two set collections is buggy in that
+      // it will fail if the ordering of elements is not equal (we just want to compare contents)
+      Assert.assertTrue(groupedMappings.get(realm).containsAll(keys));
+      Assert.assertTrue(keys.containsAll(groupedMappings.get(realm)));
+    });
+  }
+
+  /**
+   * Test that the static methods in HttpRoutingDataReader returns consistent results even though MSDS's data have been updated.
+   */
+  @Test(dependsOnMethods = "testGetMetadataStoreRoutingData")
+  public void testStaticMapping() throws IOException, InvalidRoutingDataException {
+    // Modify routing data
+    String newRealm = "newRealm";
+    _testRawRoutingData.put(newRealm, ImmutableSet.of("/newKey"));
+
+    // Kill MSDS and restart with a new mapping
+    _msdsServer.stopServer();
+    _msdsServer =
+        new MockMetadataStoreDirectoryServer(_host, _port, _namespace, _testRawRoutingData);
+    _msdsServer.startServer();
+
+    // HttpRoutingDataReader should still return old data because it's static
+    // Make sure the results don't contain the new realm
+    Map<String, List<String>> rawRoutingData = HttpRoutingDataReader.getRawRoutingData();
+    Assert.assertFalse(rawRoutingData.containsKey(newRealm));
```

Comment:
If you say "should still return old data because it's static", I think it is important to check we could still check the old data is actually returned. assert `rawRoutingData` equals to `_rawRoutingData`?

Reply:
I could do that check too.